libraries/AP_Math/SCurve.cpp

```diff
@@ -563,9 +590,47 @@ void SCurve::move_from_time_pos_vel_accel(float time_now, Vector3p &pos, Vector3
     float scurve_P1 = 0.0f;
     float scurve_V1 = 0.0f, scurve_A1 = 0.0f, scurve_J1 = 0.0f;
     get_jerk_accel_vel_pos_at_time(time_now, scurve_J1, scurve_A1, scurve_V1, scurve_P1);
-    pos += delta_unit.topostype() * (postype_t)scurve_P1;
-    vel += delta_unit * scurve_V1;
-    accel += delta_unit * scurve_A1;
+    project_scurve_onto_track(scurve_A1, scurve_V1, scurve_P1, pos, vel, accel);
+}
+
+// project the straight-line S-curve motion profile onto the active track segment
+// converts scalar S-curve kinematics (A1, V1, P1) into 3D position, velocity and acceleration
+// along either a circular arc or straight segment
+void SCurve::project_scurve_onto_track(float scurve_A1, float scurve_V1, float scurve_P1, Vector3p &pos, Vector3f &vel, Vector3f &accel)
+{
+    if (is_zero(seg_length)) {
+        return;
+    }
+
+    if (is_arc_segment) {
+        Vector2f center_to_pos_ne = -arc.center_ne;
+        const float turn_dir = is_negative(arc.angle_rad) ? -1.0f : 1.0f; // -1 for CCW, 1 for CW 
+        // rotate by progress along the arc
+        center_to_pos_ne.rotate(turn_dir * fabsf(arc.angle_rad) * (scurve_P1 / seg_length));
+
+        // position
+        const float dz_ds = seg_delta.z / seg_length;
+        Vector3f delta_pos(arc.center_ne + center_to_pos_ne, scurve_P1 * dz_ds);
+        pos += delta_pos.topostype();
+
+        // direction unit (tangent + vertical slope)
+        Vector2f arc_tangent_ne = Vector2f(-center_to_pos_ne.y, center_to_pos_ne.x) * turn_dir;
+        arc_tangent_ne /= arc.radius_ne;
+        Vector3f path_unit(arc_tangent_ne.x, arc_tangent_ne.y, dz_ds);
+        path_unit.normalize();
+
+        // velocity & tangential accel
+        vel   += path_unit * scurve_V1;
+        accel += path_unit * scurve_A1;
+
+        // centripetal accel
+        accel.xy() -=  center_to_pos_ne * sq(scurve_V1 / arc.radius_ne);
+    } else {
```

Comment:
@peterbarker would probably prefer if an early return was added a couple of lines higher and then remove this "else".